fastlane/lib/fastlane/fastfile_parser.rb

```diff
@@ -0,0 +1,324 @@
+        end
+        table_rows << [emoji, level + " ".white, linenr.to_s + " ".white, wrap_string(status, 100)]
+      end
+
+      if table_rows.length <= 0
+        return nil
+      end
+
+      table = Terminal::Table.new(title: "Fastfile Validation Result (#{@dirname})".green, headings: ["#", "State", "File/Line#", "Notice"]) do |t|
+        table_rows.each do |e|
+          t << e
+
+          t << :separator
+        end
+      end
+      return table
+    end
+
+    def find(method_name)
+      recursive_search_ast(@ast, method_name)
+      return @method_source
+    end
+
+    private
+
+    def parse(data)
+      Parser::CurrentRuby.parse(data)
+    rescue
+      return nil
+    end
+
+    # from runner.rb -> should be in FastlaneCore or somewhere shared
+    def find_alias(action_name)
+      Actions.alias_actions.each do |key, v|
+        next unless Actions.alias_actions[key]
+        next unless Actions.alias_actions[key].include?(action_name)
+        return key
+      end
+      nil
+    end
+
+    def recursive_analyze(ast)
+      if ast.nil?
+        UI.error("Parse error")
+        return nil
+      end
+      ast.children.each do |child|
+        next unless child.class.to_s == "Parser::AST::Node"
+
+        if (child.type.to_s == "send") and (child.children[0].to_s == "" && child.children[1].to_s == "lane")
+          @line_number = "#{@filename}:#{child.loc.expression.line}"
+          lane_name = child.children[2].children.first
+          lanes << lane_name
+          if Fastlane::Actions.action_class_ref(lane_name)
+            lines << { state: :info, line: @line_number, msg: "Name of the lane `#{lane_name}` already taken by action `#{lane_name}`" }
+          end
+        end
+        if (child.type.to_s == "send") and (child.children[0].to_s == "" && (Fastlane::Actions.action_class_ref(child.children[1].to_s) || find_alias(child.children[1].to_s)))
+          src_code = child.loc.expression.source
+          src_code.sub!(child.children[1].to_s, "fake_action")
+          @line_number = "#{@filename}:#{child.loc.expression.line}"
+
+          # matches = src_code.gsub!(/#\{.*\}/) do |sym|
+          #  self.class.secrets << sym if !self.class.secrets.include?(sym)
+          #  "########"
+          # end
+          copy_code = src_code.clone
+          @action_vars = []
+          src_code.scan(/#\{.*?\}/m) do |mtch|
+            # Remove #{} vars - so that there are now accidentalliy replaced ones
+            @action_vars << mtch unless self.class.secrets.include?(mtch)
+            # copy_code.gsub!(mtch.first, "'#######'")
+          end
+          src_code = copy_code
+          @original_action = child.children[1].to_s
+          dropper = '
+
+          '
+          UI.important(src_code) if $verbose
+          begin
+            Dir.chdir(@dirname) do
+              UI.important "CHDIR to: #{@dirname}" if $verbose
+              result = eval(dropper + src_code) # rubocop:disable Lint/Eval
```

Comment:
see last comment, mostly because of the need to eval the source code returned by the parser.